pkg/localstorev2/netstore.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package localstore
+
+import (
+	"context"
+	"errors"
+
+	"github.com/ethersphere/bee/pkg/pusher"
+	storage "github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	directUploadWorkers = 16
+)
+
+var errDBStopped = errors.New("db stopped")
+
+// DirectUpload is the implementation of the NetStore.DirectUpload method.
+func (db *DB) DirectUpload() PutterSession {
+	workers := make(chan struct{}, directUploadWorkers)
+	// egCtx will allow early exit of Put operations if we have
+	// already encountered error.
+	eg, egCtx := errgroup.WithContext(context.Background())
+
+	return &putterSessionImpl{
+		Putter: storage.PutterFunc(func(ctx context.Context, ch swarm.Chunk) error {
+			workers <- struct{}{}
+			eg.Go(func() error {
+				defer func() { <-workers }()
+
+				op := &pusher.Op{Chunk: ch, Err: make(chan error, 1), Direct: true}
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-egCtx.Done():
+					return egCtx.Err()
+				case <-db.quit:
+					return errDBStopped
+				case db.pusherFeed <- op:
+					select {
+					case <-ctx.Done():
+						return ctx.Err()
+					case <-egCtx.Done():
+						return egCtx.Err()
+					case <-db.quit:
+						return errDBStopped
+					case err := <-op.Err:
+						return err
+					}
+				}
+			})
+			return nil
+		}),
+		done:    func(_ swarm.Address) error { return eg.Wait() },
+		cleanup: func() error { _ = eg.Wait(); return nil },
+	}
+}
+
+// Download is the implementation of the NetStore.Download method.
+func (db *DB) Download(cache bool) storage.Getter {
+	return storage.GetterFunc(func(ctx context.Context, address swarm.Address) (swarm.Chunk, error) {
+		ch, err := db.Lookup().Get(ctx, address)
+		switch {
+		case err == nil:
+			return ch, nil
+		case errors.Is(err, storage.ErrNotFound):
+			if db.retrieval != nil {
```

Comment:
what happens if this is nil? can this ever be nil?

Reply:
If the retrieval service is not set it will just return the ErrNotFound. Right now the way to provide it is in the Options. We can have some sort of validate function on the Options which checks for necessary fields and returns error if they are not set. But this could be a valid case for eg in devmode where we dont have network retrieval.